x-pack/plugins/task_manager/lib/get_template_version.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { padLeft } from 'lodash';
+
+/*
+ * The logic for ID is: XXYYZZAA, where XX is major version, YY is minor
+ * version, ZZ is revision, and AA is alpha/beta/rc indicator.
+ *
+ * AA values below 25 are for alpha builder (since 5.0), and above 25 and below
+ * 50 are beta builds, and below 99 are RC builds, with 99 indicating a release
+ * the (internal) format of the id is there so we can easily do after/before
+ * checks on the id
+ *
+ * Note: the conversion method is carried over from Elasticsearch:
+ * https://github.com/elastic/elasticsearch/blob/de962b2/server/src/main/java/org/elasticsearch/Version.java
+ */
+export function getTemplateVersion(versionStr: string): number {
+  // break up the string parts
```

Comment:
I tested this, and it doesn't seem to handle some edge cases.

The following expression evaluates to true, but should be false (e.g. 1.32 is greater than 1.3).

`getTemplateVersion('1.32.0') < getTemplateVersion('1.3.223423')`

Is there a reason we aren't using the [semver](https://www.npmjs.com/package/semver) library to handle semvers? It's what we're using in migrations, for example.

Reply:
He matched the version behavior that Elasticsearch uses, which simply doesn't allow for 3-digit, or greater, minor and patch releases. If we get into a situation where we have Kibana 6.2.100, then I think we have other issues.

We could add a test that pulls in the packaged version and asserts that it isn't >= 99 so that we can be aware that the next minor / patch release will have that issue though.

Reply:
He matched the version behavior that Elasticsearch uses, which simply doesn't allow for 3-digit, or greater, minor and patch releases. If we get into a situation where we have Kibana 6.2.100, then I think we have other issues.

We could add a test that pulls in the packaged version and asserts that it isn't >= 99 so that we can be aware that the next minor / patch release will have that issue though.

Reply:
This thread is relevant to a call we had with @njd5475 and @chrisdavies and others

1. What actually "upgrades" existing documents in the task manager index? 
  Since that isn't a concern to worry about for this version, I haven't been addressing that very much. I think this PR will create a [logical space in the code](https://github.com/elastic/kibana/pull/28540/files#diff-2d129f8a9b099164fc572879265de210R143) where upgrade logic will go in the future.
2. Use semver? Or integer?
  In this PR, I'm setting up a way to define what version of Kibana a document will be compatible with, and doing that with the `kibana.apiVersion` field. That lets me [create a filter in the search query to ES](https://github.com/elastic/kibana/pull/28540/files#diff-2d129f8a9b099164fc572879265de210R298). However, if the documents are going to be compatible with the Kibana Migrations service, their versioning will be done with semver, and the version should not necessarily update with each release. Also, if the documents were stored using the Saved Objects client, this would happen automatically.

If I understand things correctly, Kibana Migrations and Saved Objects client mean we don't need to worry about BWC in our own code? That would be great! Although, for now there seem to be no features that use those capabilities other than core Kibana for the `.kibana` index. Given that, I think the more hands-on versioning that I'm doing here is OK for now (Task Manager has no consumers yet, and is a beta service). In the future, maintainers can go over `TaskStore` and fashion it to use the Saved Objects client.

Does this sound ok?

Reply:
cc @cwurm, was on the call and had some thoughts about Saved Object client as well

Reply:
Not me :)

Reply:
sorry - I meant @weltenwort